Packages/io.chainsafe.web3-unity/Runtime/Common/Scripts/LoginProvider.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Threading.Tasks;
+using ChainSafe.Gaming.UnityPackage.Common;
+using UnityEngine;
+
+namespace Scenes
+{
+    /// <summary>
+    /// A concrete implementation of <see cref="ILoginProvider"/>.
+    /// </summary>
+    public abstract class LoginProvider : MonoBehaviour, ILoginProvider
+    {
+        [SerializeField] private string gelatoApiKey = "";
+
+        [SerializeField] private ErrorPopup errorPopup;
+
+        public string GelatoApiKey => gelatoApiKey;
+        public IWeb3BuilderServiceAdapter[] Web3BuilderServiceAdapters { get; private set; }
+        public IWeb3InitializedHandler[] Web3InitializedHandlers { get; private set; }
+
+        private void Start()
+        {
+            Initialize();
+        }
+
+        /// <summary>
+        /// Initializes Login providers.
+        /// </summary>
+        protected virtual void Initialize()
+        {
+            Web3BuilderServiceAdapters = GetComponents<IWeb3BuilderServiceAdapter>();
+            
+            Web3InitializedHandlers = GetComponents<IWeb3InitializedHandler>();
+        }
+
+        /// <summary>
+        /// Try to Login and displays error and throws exception on a failed attempt.
+        /// </summary>
+        public async Task TryLogin()
+        {
+            try
+            {
+                await (this as ILoginProvider).Login();
```

Comment:
was this cast necessary?

Reply:
Yes, Default interface implementations work similarly to explicit implementations: they can only be called through the interface type, not the implementing type.